In [3]:
# Data processing
import pandas as pd
import numpy as np
import scipy.stats

# Visualization
import seaborn as sb

# Similarity
from sklearn.metrics.pairwise import cosine_similarity

# Other
import math
import random
import sklearn
from nltk.corpus import stopwords
from scipy.sparse import csr_matrix
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from scipy.sparse.linalg import svds
from sklearn.preprocessing import MinMaxScaler
import matplotlib.pyplot as plt

In [4]:
ratings = pd.read_csv("../archive/rating_complete.csv")
ratings

user_id  anime_id  rating
0               0       430       9
1               0      1004       5
2               0      3010       7
3               0       570       7
4               0      2762       9
...           ...       ...     ...
57633273   353404       502       8
57633274   353404       987       4
57633275   353404       225       8
57633276   353404       243       7
57633277   353404       896       7

[57633278 rows x 3 columns]

In [5]:
anime_df = pd.read_csv("../archive/anime.csv")
anime_df["anime_id"] = anime_df["MAL_ID"]
anime_df

MAL_ID                             Name    Score  \
0           1                     Cowboy Bebop     8.78   
1           5  Cowboy Bebop: Tengoku no Tobira     8.39   
2           6                           Trigun     8.24   
3           7               Witch Hunter Robin     7.27   
4           8                   Bouken Ou Beet     6.98   
...       ...                              ...      ...   
17557   48481  Daomu Biji Zhi Qinling Shen Shu  Unknown   
17558   48483                     Mieruko-chan  Unknown   
17559   48488  Higurashi no Naku Koro ni Sotsu  Unknown   
17560   48491      Yama no Susume: Next Summit  Unknown   
17561   48492                    Scarlet Nexus  Unknown   

                                                  Genres  \
0        Action, Adventure, Comedy, Drama, Sci-Fi, Space   
1                  Action, Drama, Mystery, Sci-Fi, Space   
2      Action, Sci-Fi, Adventure, Comedy, Drama, Shounen   
3      Action, Mystery, Police, Supernatural, Drama, ...   
4              Adventure, Fantasy, Shounen, Supernatural   
...                                                  ...   
17557                   Adventure, Mystery, Supernatural   
17558                       Comedy, Horror, Supernatural   
17559  Mystery, Dementia, Horror, Psychological, Supe...   
17560                   Adventure, Slice of Life, Comedy   
17561                                    Action, Fantasy   

                          English name                     Japanese name  \
0                         Cowboy Bebop                         カウボーイビバップ   
1               Cowboy Bebop:The Movie                    カウボーイビバップ 天国の扉   
2                               Trigun                             トライガン   
3                   Witch Hunter Robin  Witch Hunter ROBIN (ウイッチハンターロビン)   
4               Beet the Vandel Buster                            冒険王ビィト   
...                                ...                               ...   
17557                          Unknown                         盗墓笔记之秦岭神树   
17558                          Unknown                           見える子ちゃん   
17559  Higurashi:When They Cry – SOTSU                        ひぐらしのなく頃に卒   
17560                          Unknown                ヤマノススメ Next Summit   
17561                          Unknown                     SCARLET NEXUS   

        Type Episodes                         Aired    Premiered  ...  \
0         TV       26   Apr 3, 1998 to Apr 24, 1999  Spring 1998  ...   
1      Movie        1                   Sep 1, 2001      Unknown  ...   
2         TV       26   Apr 1, 1998 to Sep 30, 1998  Spring 1998  ...   
3         TV       26   Jul 2, 2002 to Dec 24, 2002  Summer 2002  ...   
4         TV       52  Sep 30, 2004 to Sep 29, 2005    Fall 2004  ...   
...      ...      ...                           ...          ...  ...   
17557    ONA  Unknown              Apr 4, 2021 to ?      Unknown  ...   
17558     TV  Unknown                     2021 to ?      Unknown  ...   
17559     TV  Unknown                Jul, 2021 to ?  Summer 2021  ...   
17560     TV  Unknown                       Unknown      Unknown  ...   
17561     TV  Unknown                Jul, 2021 to ?  Summer 2021  ...   

        Score-9   Score-8  Score-7  Score-6  Score-5  Score-4  Score-3  \
0      182126.0  131625.0  62330.0  20688.0   8904.0   3184.0   1357.0   
1       49201.0   49505.0  22632.0   5805.0   1877.0    577.0    221.0   
2       75651.0   86142.0  49432.0  15376.0   5838.0   1965.0    664.0   
3        4806.0   10128.0  11618.0   5709.0   2920.0   1083.0    353.0   
4         529.0    1242.0   1713.0   1068.0    634.0    265.0     83.0   
...         ...       ...      ...      ...      ...      ...      ...   
17557   Unknown   Unknown      1.0  Unknown  Unknown  Unknown  Unknown   
17558   Unknown   Unknown  Unknown  Unknown  Unknown  Unknown  Unknown   
17559   Unknown   Unknown  Unknown  Unknown  Unknown  Unknown  Unknown   
17560   Unknown   Unknown  Unknown  Unknown  Un

In [6]:
df = pd.merge(ratings, anime_df[['anime_id','Name', 'Genres']], on="anime_id", how="inner")
df

user_id  anime_id  rating  \
0               0       430       9   
1               6       430       8   
2              18       430      10   
3              19       430       8   
4              33       430       4   
...           ...       ...     ...   
57633273   315549     38853       1   
57633274   350024     38853      10   
57633275   311855     39435       6   
57633276   334473     35746       4   
57633277   347134     39627       5   

                                                     Name  \
0         Fullmetal Alchemist: The Conqueror of Shamballa   
1         Fullmetal Alchemist: The Conqueror of Shamballa   
2         Fullmetal Alchemist: The Conqueror of Shamballa   
3         Fullmetal Alchemist: The Conqueror of Shamballa   
4         Fullmetal Alchemist: The Conqueror of Shamballa   
...                                                   ...   
57633273                                           Ex-Arm   
57633274                                           Ex-Arm   
57633275          Oshiri Tantei Movie 1: Curry Naru Jiken   
57633276                                         6 Lovers   
57633277                        Nan-chan no Koutsuu Anzen   

                                                     Genres  
0         Military, Comedy, Historical, Drama, Fantasy, ...  
1         Military, Comedy, Historical, Drama, Fantasy, ...  
2         Military, Comedy, Historical, Drama, Fantasy, ...  
3         Military, Comedy, Historical, Drama, Fantasy, ...  
4         Military, Comedy, Historical, Drama, Fantasy, ...  
...                                                     ...  
57633273                      Action, Sci-Fi, Ecchi, Seinen  
57633274                      Action, Sci-Fi, Ecchi, Seinen  
57633275                     Mystery, Comedy, Kids, Fantasy  
57633276  Comedy, Drama, Romance, School, Shounen Ai, Sl...  
57633277                                               Kids  

[57633278 rows x 5 columns]

# Data Preparation
* 57,633,278 rows are to many entries to compute, let's reduce the dataset

In [7]:
# User ratings
print(df["user_id"].value_counts().mean())
df["user_id"].value_counts()

185.87842313882197


189037    15455
162615    14864
68042     13462
283786    12778
259790     9996
          ...  
136422        1
213994        1
196985        1
190834        1
206653        1
Name: user_id, Length: 310059, dtype: int64

In [8]:
# Anime ratings
print(df["anime_id"].value_counts().mean())
df["anime_id"].value_counts()

3415.912636320531


1535     182375
16498    169794
11757    161192
6547     141127
30276    138924
          ...  
39686         1
39685         1
35153         1
40594         1
39627         1
Name: anime_id, Length: 16872, dtype: int64

In [9]:
# Filter out users with less than 150 ratings and animes with less than 3500 ratings
df = df.groupby("user_id").filter(lambda x: len(x) > 150) # 310059
df = df.groupby("anime_id").filter(lambda x: len(x) > 6000) # 16872

In [10]:
df.info()
# User ratings
print("USERS \n")
print(df["user_id"].value_counts().mean())
print(df["user_id"].value_counts())
print("\n")

print("ANIMES \n")
# Anime ratings
print(df["anime_id"].value_counts().mean())
print(df["anime_id"].value_counts())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 35884969 entries, 1 to 55060794
Data columns (total 5 columns):
 #   Column    Dtype 
---  ------    ----- 
 0   user_id   int64 
 1   anime_id  int64 
 2   rating    int64 
 3   Name      object
 4   Genres    object
dtypes: int64(3), object(2)
memory usage: 1.6+ GB
USERS 

287.47071216854926
277841    1945
291207    1934
189037    1933
283786    1918
68042     1915
          ... 
142535       1
155668       1
194630       1
200466       1
120222       1
Name: user_id, Length: 124830, dtype: int64


ANIMES 

18151.223571067272
16498    97061
11757    95262
1535     94175
6547     89411
4224     86389
         ...  
34148     6035
7655      6021
369       6018
34544     6016
8634      6011
Name: anime_id, Length: 1977, dtype: int64


In [11]:
# We are left still with 38,616,425 entries even after filtering out users and animes
# so let's sample the remaining dataset so we get a smaller one

df_sample = df.sample(frac=0.001)
df_sample # 34622 entries seems more logical for first runs so we can re-run faster

user_id  anime_id  rating  \
3202116    308439     33486       9   
29249435   323066      9260       7   
36857423   320998     11313      10   
10965101   344198        30      10   
39539247    96992        99       8   
...           ...       ...     ...   
35271665   221664      1827       8   
32178863   114959      9790       8   
12997136    59244     15583       8   
24089064   301612     10162       9   
41028270   114260     34383       4   

                                                 Name  \
3202116              Boku no Hero Academia 2nd Season   
29249435               Kizumonogatari I: Tekketsu-hen   
36857423         Kimi no Iru Machi: Tasogare Kousaten   
10965101                      Neon Genesis Evangelion   
39539247                                    Mai-Otome   
...                                               ...   
35271665                           Seirei no Moribito   
32178863  Sora no Otoshimono: Tokeijikake no Angeloid   
12997136                                  Date A Live   
24089064                                   Usagi Drop   
41028270                                Netsuzou TRap   

                                                     Genres  
3202116        Action, Comedy, Super Power, School, Shounen  
29249435                                            Vampire  
36857423                    Drama, Romance, School, Shounen  
10965101  Action, Sci-Fi, Dementia, Psychological, Drama...  
39539247                      Comedy, Drama, Fantasy, Magic  
...                                                     ...  
35271665             Action, Adventure, Historical, Fantasy  
32178863  Comedy, Drama, Ecchi, Harem, Romance, Sci-Fi, ...  
12997136      Sci-Fi, Harem, Comedy, Romance, Mecha, School  
24089064                               Slice of Life, Josei  
41028270                          Drama, Romance, Shoujo Ai  

[35885 rows x 5 columns]

# Evaluation
Divide df into training and test set (80 - 20)

In [12]:
train_df, test_df = train_test_split(df_sample.drop(columns={"Name","Genres"}),
                                #    stratify=df_sample['user_id'], 
                                   test_size=0.20,
                                   random_state=42)

print('# ratings on Train set: %d' % len(train_df))
print('# ratings on Test set: %d' % len(test_df))
train_df

# ratings on Train set: 28708
# ratings on Test set: 7177


user_id  anime_id  rating
25972329   243115     20689       6
23085066   132614      3901       7
36213076   279140     20057      10
14402194   328397      9989       5
47993558   260980     40595       8
...           ...       ...     ...
37666490   175795     30206       8
33231436    63509     30654       9
8612398    101031      8074       8
44137190   301820     21855       7
23976407   241281      1840       7

[28708 rows x 3 columns]

In Recommender Systems, there are a set metrics commonly used for evaluation. We chose to work with Top-N accuracy metrics, which evaluates the accuracy of the top recommendations provided to a user, comparing to the items the user has actually interacted in test set.
This evaluation method works as follows:

* For each user
    * For each item the user has interacted in test set
        * Sample 100 other items the user has never interacted.
            
            Ps: Here we naively assume those non interacted items are not relevant to the user, which might not be true, as the user may simply not be aware of those not interacted items. But let's keep this assumption.
        * Ask the recommender model to produce a ranked list of recommended items, from a set composed one interacted item and the 100 non-interacted ("non-relevant!) items
        * Compute the Top-N accuracy metrics for this user and interacted item from the recommendations ranked list
* Aggregate the global Top-N accuracy metrics

In [13]:
#Indexing by personId to speed up the searches during evaluation
indexed_df = df_sample.set_index('user_id')
train_indexed_df = train_df.set_index('user_id')
test_indexed_df = test_df.set_index('user_id')

In [14]:
def get_items_interacted(person_id, interactions_df):
    # Get the user's data and merge in the movie information.

    if person_id in interactions_df.index :
        interacted_items = interactions_df.loc[person_id]['anime_id']
    else:
        return []
    
    # print(interacted_items.head(1))
    return set(interacted_items if type(interacted_items) == pd.Series else [interacted_items])

In [15]:
#Top-N accuracy metrics consts
EVAL_RANDOM_SAMPLE_NON_INTERACTED_ITEMS = 100

class ModelEvaluator:


    def get_not_interacted_items_sample(self, person_id, sample_size, seed=42):
        interacted_items = get_items_interacted(person_id, indexed_df.drop(columns={"Name","Genres"}))
        all_items = set(indexed_df['anime_id'])
        non_interacted_items = all_items - interacted_items
        
        random.seed(seed)
        non_interacted_items_sample = random.sample(non_interacted_items, sample_size)
        return set(non_interacted_items_sample)

    def _verify_hit_top_n(self, item_id, recommended_items, topn):        
            try:
                index = next(i for i, c in enumerate(recommended_items) if c == item_id)
            except:
                index = -1
            hit = int(index in range(0, topn))
            return hit, index

    def evaluate_model_for_user(self, model, person_id):
        #Getting the items in test set
        interacted_values_testset = test_indexed_df.loc[person_id]
        if type(interacted_values_testset['anime_id']) == pd.Series:
            person_interacted_items_testset = set(interacted_values_testset['anime_id'])
        else:
            person_interacted_items_testset = set([int(interacted_values_testset['anime_id'])])  
        interacted_items_count_testset = len(person_interacted_items_testset) 

        #Getting a ranked recommendation list from a model for a given user
        person_recs_df = model.recommend_items(person_id, 
                                               items_to_ignore=get_items_interacted(person_id, 
                                                                                    train_indexed_df), 
                                               topn=10000000000)
        hits_at_5_count = 0
        hits_at_10_count = 0

        #For each item the user has interacted in test set
        for item_id in person_interacted_items_testset:
            #Getting a random sample (100) items the user has not interacted 
            #(to represent items that are assumed to be no relevant to the user)
            non_interacted_items_sample = self.get_not_interacted_items_sample(person_id, 
                                                                          sample_size=EVAL_RANDOM_SAMPLE_NON_INTERACTED_ITEMS, 
                                                                          seed=item_id%(2**32))
            #Combining the current interacted item with the 100 random items
            items_to_filter_recs = non_interacted_items_sample.union(set([item_id]))

            #Filtering only recommendations that are either the interacted item or from a random sample of 100 non-interacted items
            valid_recs_df = person_recs_df[person_recs_df['anime_id'].isin(items_to_filter_recs)]                    
            valid_recs = valid_recs_df['anime_id'].values
            #Verifying if the current interacted item is among the Top-N recommended items
            hit_at_5, index_at_5 = self._verify_hit_top_n(item_id, valid_recs, 5)
            hits_at_5_count += hit_at_5
            hit_at_10, index_at_10 = self._verify_hit_top_n(item_id, valid_recs, 10)
            hits_at_10_count += hit_at_10

        #Recall is the rate of the interacted items that are ranked among the Top-N recommended items, 
        #when mixed with a set of non-relevant items
        recall_at_5 = hits_at_5_count / float(interacted_items_count_testset)
        recall_at_10 = hits_at_10_count / float(interacted_items_count_testset)
        person_metrics = {'hits@5_count':hits_at_5_count, 
                          'hits@10_count':hits_at_10_count, 
                          'interacted_count': interacted_items_count_testset,
                          'recall@5': recall_at_5,
                          'recall@10': recall_at_10}
        # ADEDEDEDEDD
        # print(person_metrics)
        return person_metrics

    def evaluate_model(self, model):
        #print('Running evaluation for users')
        people_metrics = []
        for idx, person_id in enumerate(list(test_indexed_df.index.unique().values)):
            if idx % 100 == 0 and idx > 0:
               print('%d users processed' % idx)
            person_metrics = self.evaluate_model_for_user(model, person_id)  
            person_metrics['user_id'] = person_id
            people_metrics.append(person_metrics)
            # if(idx > 15):
            #     break
        print('%d users processed' % idx)

        detailed_results_df = pd.DataFrame(people_metrics) \
                            .sort_values('interacted_count', ascending=False)
            
        global_recall_at_5 = detailed_results_df['hits@5_count'].sum() / float(detailed_results_df['interacted_count'].sum())
        global_recall_at_10 = detailed_results_df['hits@10_count'].sum() / float(detailed_results_df['interacted_count'].sum())
        
        global_metrics = {'modelName': model.get_model_name(),
                          'recall@5': global_recall_at_5,
                          'recall@10': global_recall_at_10}    
        return global_metrics, detailed_results_df
    
model_evaluator = ModelEvaluator()    

## Popularity model

In [16]:
#Computes the most popular items
item_popularity_df = indexed_df.groupby('anime_id')['rating'].sum().sort_values(ascending=False).reset_index()
item_popularity_df.head(10)

anime_id  rating
0      1535     895
1     16498     861
2     19815     787
3     32281     785
4     20507     768
5      9253     743
6      5114     736
7     30276     735
8      4224     734
9      8074     715

In [17]:
class PopularityRecommender:
    
    MODEL_NAME = 'Popularity'
    
    def __init__(self, popularity_df, items_df=None):
        self.popularity_df = popularity_df
        self.items_df = items_df
        
    def get_model_name(self):
        return self.MODEL_NAME
        
    def recommend_items(self, user_id, items_to_ignore=[], topn=10, verbose=False):
        # Recommend the more popular items that the user hasn't seen yet.
        recommendations_df = self.popularity_df[~self.popularity_df['anime_id'].isin(items_to_ignore)] \
                               .sort_values('rating', ascending = False) \
                               .head(topn)

        if verbose:
            if self.items_df is None:
                raise Exception('"items_df" is required in verbose mode')
    
            recommendations_df = recommendations_df.merge(self.items_df, how = 'left', 
                                                          left_on = 'anime_id', 
                                                          right_on = 'anime_id')[['rating']]


        return recommendations_df
    
popularity_model = PopularityRecommender(item_popularity_df, indexed_df)

In [18]:
# import warnings
# warnings.filterwarnings('ignore')
print('Evaluating Popularity recommendation model... (', len(train_indexed_df.value_counts()), ' users )')
pop_global_metrics, pop_detailed_results_df = model_evaluator.evaluate_model(popularity_model)
print('\nGlobal metrics:\n%s' % pop_global_metrics)
pop_detailed_results_df.head(10)

Evaluating Popularity recommendation model... ( 9256  users )


TypeError: Population must be a sequence.  For dicts or sets, use sorted(d).